src/pgzero/joystick.py

```diff
@@ -0,0 +1,577 @@
+    RX = 3
+    RY = 4
+    RT = 5
+
+
+BTN_NAMES = ("face_up", "face_down", "face_left", "face_right", "dpad_up",
+             "dpad_down", "dpad_left", "dpad_right", "shoulder_left",
+             "shoulder_right", "push_left", "push_right", "center_left",
+             "center_middle", "center_right")
+PressedBtns = namedtuple("PressedBtns", BTN_NAMES)
+
+
+class Joystick:
+    """PGZero wrapper class for joystick use."""
+
+    def __init__(self, stick):
+        self._stick = stick
+        name = self._stick.get_name()
+        # We use the system name of the controller to determine
+        # the correct mappings to use.
+        if "Series" in name:
+            self._type = "XSeries"
+            self._btn_map = XSeriesBtns
+            self._axis_map = XSeriesAxis
+        elif "Xbox" in name:
+            self._type = "Xbox"
+            self._btn_map = X360Btns
+            self._axis_map = X360Axis
+        elif "PS4" in name:
+            self._type = "PS4"
+            self._btn_map = PS4Btns
+            self._axis_map = PS4Axis
+        elif "Sony Interactive" in name:
+            self._type = "PS5"
+            self._btn_map = PS5Btns
+            self._btn_map = PS5Axis
+        elif "Switch" in name:
+            self._type = "Switch"
+            self._btn_map = SwitchBtns
+            self._axis_map = SwitchAxis
+        else:
+            self._type = "UNKNOWN"
+            # If the type is unknown, we fall back to Xbox
+            # mappings since these are most common among
+            # generic controller layouts.
+            self._btn_map = X360Btns
+            self._axis_map = X360Axis
+
+        self._pressed = [False] * self._btn_map.NUM
+        self._axis = [0] * self._axis_map.NUM
+        # Since triggers don't start centered but rather unpressed,
+        # their starting values are -1.
+        self._axis[self._axis_map.RT] = -1
+        self._axis[self._axis_map.LT] = -1
+
+        # These variables are used to give a lockout time for physical
+        # HAT button simulation to prevent ghost inputs.
+        self._DU_open = True
+        self._DD_open = True
+        self._DL_open = True
+        self._DR_open = True
+
+    # Used with clock.schedule to prevent ghost button presses via timeouts.
+    def _unlock_DU(self):
+        self._DU_open = True
+
+    def _unlock_DD(self):
+        self._DD_open = True
+
+    def _unlock_DL(self):
+        self._DL_open = True
+
+    def _unlock_DR(self):
+        self._DR_open = True
+
+    @property
+    def instance_id(self):
+        """Returns the instance id of the controller."""
+        return self._stick.get_instance_id()
+
+    @property
+    def pressed(self):
+        """Returns a named tuple with the current pressed state
+        of all buttons on the controller."""
+        btns = iter(self._btn_map)
+        next(btns)
+        states = []
+        for b in btns:
+            states.append(self._pressed[b])
+        return PressedBtns(*states)
```

Comment:
I don't like this API of copying states into a tuple just to access a couple of values, it's really unperformant. Pygame does it for the keyboard but that's quite odd. I think the attributes of this class are quite satisfactory.